### PR TITLE
perf(encoding): add estimate size for value encoding

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

add function to estimate the value encoding size of the datum
prepare for https://github.com/risingwavelabs/risingwave/issues/8683

## Checklist For Contributors

- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/src/util/value_encoding/mod.rs [hunk 1]
```diff
@@ -225,6 +233,28 @@ fn serialize_scalar(value: ScalarRefImpl<'_>, buf: &mut impl BufMut) {
...
+        ScalarRefImpl::Decimal(_) => estimate_encoded_decimal_size(),
+        ScalarRefImpl::Interval(_) => estimate_encoded_interval_size(),
+        ScalarRefImpl::NaiveDate(_) => estimate_encoded_naivedate_size(),
+        ScalarRefImpl::NaiveDateTime(_) => estimate_encoded_naivedatetime_size(),
+        ScalarRefImpl::NaiveTime(_) => estimate_encoded_naivetime_size(),
+        ScalarRefImpl::Jsonb(_) => 8,
+        ScalarRefImpl::Struct(s) => estimate_encoded_struct_size(s),
+        ScalarRefImpl::List(v) => estimate_encoded_list_size(v),
+    }
+}
```
Comment (st1page) [inline src/common/src/util/value_encoding/mod.rs:256]: maybe we can refactor it with macro or other techs in future PRs.

**Review by wangrunji0408 — APPROVED** [review]

**Comment by TennyZhuang** [comment]
1. Why it’s estimated_size? IIUC it’s exact.
2. It’s better to have fuzz test here, but It may be a little complicated. At least, you should write UT for every type to ensure the format will not be broken silently.

**Review by TennyZhuang — CHANGES_REQUESTED** [review]
Please add test

**Comment by TennyZhuang** [comment]
You can write some datums, encode them, and compare with your pre-calculated result.

File: src/common/src/util/value_encoding/mod.rs [hunk 2]
```diff
@@ -379,3 +439,65 @@ fn deserialize_decimal(data: &mut impl Buf) -> Result<Decimal> {
...
+        test_estimate_serialize_scalar_size(ScalarImpl::NaiveTime(
+            NaiveTimeWrapper::from_hms_uncheck(2, 3, 3),
+        ));
+        test_estimate_serialize_scalar_size(ScalarImpl::NaiveDateTime(
+            NaiveDateTimeWrapper::from_timestamp_uncheck(23333333, 2333),
+        ));
+        test_estimate_serialize_scalar_size(ScalarImpl::Interval(
+            IntervalUnit::from_month_day_usec(2, 3, 3333),
+        ));
+        // test_estimate_serialize_scalar_size(ScalarImpl::Jsonb(JsonbVal::dummy()));
```
Comment (TennyZhuang) [inline src/common/src/util/value_encoding/mod.rs:493]:
```suggestion
```

**Review by TennyZhuang — APPROVED** [review]

File: src/common/src/util/value_encoding/mod.rs [hunk 3]
```diff
@@ -183,6 +183,14 @@ pub fn serialize_datum_into(datum_ref: impl ToDatumRef, buf: &mut impl BufMut) {
     }
 }
 
+pub fn estimate_serialize_datum_size(datum_ref: impl ToDatumRef) -> usize {
+    if let Some(d) = datum_ref.to_datum_ref() {
+        1 + estimate_serialize_scalar_size(d)
+    } else {
+        1
```
Comment (kwannoel) [inline src/common/src/util/value_encoding/mod.rs:190]: I guess `1` here is to store the null bit flag?

Comment (st1page) [reply src/common/src/util/value_encoding/mod.rs:190]: Yes

File: src/common/src/util/value_encoding/mod.rs [hunk 4]
```diff
@@ -225,6 +233,28 @@ fn serialize_scalar(value: ScalarRefImpl<'_>, buf: &mut impl BufMut) {
...
+        ScalarRefImpl::Float64(_) => 8,
+        ScalarRefImpl::Utf8(v) => estimate_serialize_str_size(v.as_bytes()),
+        ScalarRefImpl::Bytea(v) => estimate_serialize_str_size(v),
+        ScalarRefImpl::Bool(_) => 1,
+        ScalarRefImpl::Decimal(_) => estimate_serialize_decimal_size(),
+        ScalarRefImpl::Interval(_) => estimate_serialize_interval_size(),
+        ScalarRefImpl::NaiveDate(_) => estimate_serialize_naivedate_size(),
+        ScalarRefImpl::NaiveDateTime(_) => estimate_serialize_naivedatetime_size(),
+        ScalarRefImpl::NaiveTime(_) => estimate_serialize_naivetime_size(),
+        ScalarRefImpl::Jsonb(_) => 8,
```
Comment (fuyufjh) [inline src/common/src/util/value_encoding/mod.rs:252]: Is this a mistake? @st1page

Comment (TennyZhuang) [reply src/common/src/util/value_encoding/mod.rs:252]: It’s heavy to pre-compute the size of jsonb, but anyway, 8 is too small and we need some comments here.

Comment (st1page) [reply src/common/src/util/value_encoding/mod.rs:252]: 🤔 I will leave an issue to bench and trade-off if it is worth estimating the jsonb's size. @st1page

**Pull request merged**